File: eren_3d/src/game_world/transform.rs
use glam::{Mat4, Quat, Vec3};

pub struct LocalTransform {
    position: Vec3,
    pivot: Vec3,
    scale: Vec3,
    rotation: Quat,
    alpha: f32,
    is_dirty: bool,
}

impl LocalTransform {
    pub fn new() -> Self {
        Self {
            position: Vec3::ZERO,
            pivot: Vec3::ZERO,
            scale: Vec3::ONE,
            rotation: Quat::IDENTITY,
            alpha: 1.0,
            is_dirty: true,
        }
    }

    pub fn position(&self) -> Vec3 {
        self.position
    }

    pub fn set_position(&mut self, value: Vec3) {
        if self.position != value {
            self.position = value;
            self.is_dirty = true;
        }
    }

    pub fn pivot(&self) -> Vec3 {
        self.pivot
    }

    pub fn set_pivot(&mut self, value: Vec3) {
        if self.pivot != value {
            self.pivot = value;
            self.is_dirty = true;
        }
    }

    pub fn scale(&self) -> Vec3 {
        self.scale
    }

    pub fn set_scale(&mut self, value: Vec3) {
        if self.scale != value {
            self.scale = value;
            self.is_dirty = true;
        }
    }

    pub fn rotation(&self) -> Quat {
        self.rotation
    }

    pub fn set_rotation(&mut self, value: Quat) {
        if self.rotation != value {
            self.rotation = value;
            self.is_dirty = true;
        }
    }

    pub fn alpha(&self) -> f32 {
        self.alpha
    }

    pub fn set_alpha(&mut self, value: f32) {
        if self.alpha != value {
            self.alpha = value;
            self.is_dirty = true;
        }
    }
}

pub struct GlobalTransform {
    matrix: Mat4,
    alpha: f32,
    is_dirty: bool,
}

impl GlobalTransform {
    pub fn new() -> Self {
        Self {
            matrix: Mat4::IDENTITY,
            alpha: 1.0,
            is_dirty: false,
        }
    }

    pub fn update(&mut self, parent: &GlobalTransform, local: &mut LocalTransform) {
        if parent.is_dirty || local.is_dirty {
            let pivot_transform = Mat4::from_translation(local.pivot)
                * Mat4::from_quat(local.rotation)
                * Mat4::from_scale(local.scale)
                * Mat4::from_translation(-local.pivot);

            let local_matrix =
                Mat4::from_translation(local.position - local.pivot) * pivot_transform;

            self.matrix = parent.matrix * local_matrix;
            self.alpha = parent.alpha * local.alpha;
            self.is_dirty = true;

            local.is_dirty = false;
        }
    }

    pub fn finalize(&mut self) {
        self.is_dirty = false;
    }

    pub fn extract(&mut self) -> (Mat4, f32) {
        self.finalize();
        (self.matrix, self.alpha)
    }
}

File: eren_ts_2d/src/transform.ts
import { mat3, vec2 } from 'gl-matrix';

export class Transform {
  private _position: vec2;
  private _pivot: vec2;
  private _scale: vec2;
  private _rotation: number;
  private _alpha: number;
  private _isDirty: boolean;

  constructor() {
    this._position = vec2.create();
    this._pivot = vec2.create();
    this._scale = vec2.create();
    this._rotation = 0;
    this._alpha = 1;
    this._isDirty = true;
  }

  public get position(): vec2 {
    return this._position;
  }

  public set position(value: vec2) {
    this._position = value;
    this._isDirty = true;
  }

  public get pivot(): vec2 {
    return this._pivot;
  }

  public set pivot(value: vec2) {
    this._pivot = value;
    this._isDirty = true;
  }

  public get scale(): vec2 {
    return this._scale;
  }

  public set scale(value: vec2) {
    this._scale = value;
    this._isDirty = true;
  }

  public get rotation(): number {
    return this._rotation;
  }

  public set rotation(value: number) {
    this._rotation = value;
    this._isDirty = true;
  }

  public get alpha(): number {
    return this._alpha;
  }

  public set alpha(value: number) {
    this._alpha = value;
    this._isDirty = true;
  }

  public isDirty(): boolean {
    return this._isDirty;
  }

  public clearDirty() {
    this._isDirty = false;
  }
}

export class GlobalTransform {
  private _matrix: mat3;
  private _alpha: number;
  private _isDirty: boolean;

  constructor() {
    this._matrix = mat3.create();
    this._alpha = 1;
    this._isDirty = false;
  }

  // 성능 최적화용: 재사용할 임시 객체
  private _t1 = mat3.create();
  private _r = mat3.create();
  private _s = mat3.create();
  private _t2 = mat3.create();
  private _pivot_transform = mat3.create();
  private _translation = mat3.create();
  private _offset = vec2.create();

  public update(parent: GlobalTransform, local: Transform) {
    if (!parent._isDirty && !local.isDirty()) return;

    const t1 = this._t1;
    const r = this._r;
    const s = this._s;
    const t2 = this._t2;
    const pivotTransform = this._pivot_transform;
    const translation = this._translation;
    const offset = this._offset;

    // pivot_transform = T(pivot) * R(rotation) * S(scale) * T(-pivot)
    mat3.fromTranslation(t1, local.pivot);
    mat3.fromRotation(r, local.rotation);
    mat3.fromScaling(s, local.scale);

    vec2.negate(offset, local.pivot);
    mat3.fromTranslation(t2, offset);

    mat3.multiply(pivotTransform, t1, r);
    mat3.multiply(pivotTransform, pivotTransform, s);
    mat3.multiply(pivotTransform, pivotTransform, t2);

    // local_matrix = T(position - pivot) * pivot_transform
    vec2.sub(offset, local.position, local.pivot);
    mat3.fromTranslation(translation, offset);
    const localMatrix = translation;

    mat3.multiply(this._matrix, parent._matrix, localMatrix);
    mat3.multiply(this._matrix, this._matrix, pivotTransform);

    this._alpha = parent._alpha * local.alpha;
    this._isDirty = true;

    local.clearDirty(); // optional setter
  }

  public finalize() {
    this._isDirty = false;
  }

  // 아래와 같은 함수는 매 프레임마다 오브젝트를 생성시키므로, 성능에 좋지 않아 사용하지 않음
  /*public extract() {
    this.finalize();
    return { matrix: this._matrix, alpha: this._alpha };
  }*/
}
